Articles/ShortArticles/PullRequestSizeConsiderations.md

```diff
@@ -0,0 +1,155 @@
+for CI in each PR, ensuring each PR is reviewed and merged in logical order and
+possibly having to apply each PR's changes to multiple branches (e.g. release candidate and main).
+
+In addition, when splitting large changes over multiple PRs, there can be some amount
+of *transition-specific* software engineering required (design and coding work devoted
+solely to holding things together *during* a major transition) which is later removed. 
+This effort is needed *only* to support the software through multiple stages of a major
+transition which would have otherwise been unnecessary. So, it represents extra work.
+
+The point is, productivity for the project *overall* requires
+*balancing* all these concerns.<sup>[19]</sup>
+In some cases strict adherence to the principle of not mixing unrelated changes in a 
+single PR may outweigh the benefits of splitting such work across multiple PRs.
+In other words, for very small bits of work (e.g. fixing a typo in a document), there may be
+a minimum size PR below which the overheads of just managing the PR process become
+the key productivity concern. Still, this should be something that is
+discussed with the team ahead of submitting the changes.
+
+The upshot is that keeping PRs as small as reasonably practical is a best practice. Investing
+time up front to decide how to split large changes into manageable PRs that can be quickly
+integrated into a large code base is as important as the design and coding work that goes
+into the bug fixes and feature enhancements the changes are proposed for to begin with.
+
+Some literature uses the notion of a *reviewable unit of work*<sup>[20]</sup>
+to separate the concept of code review from how it is practiced in any particular tool.
+A reviewable unit could be a
+PR, a commit, or a single file's patch/diff.<sup>[21]</sup>
+In fact, long before GitHub introduced pull requests in 2008 (which was
+in turn based on Git's `request-pull`<sup>[22]</sup> operation),
+seasoned software professionals engaged in the conceptually equivalent processes using a
+patchwork of approaches and tools<sup>[23]</sup>
+(or using something like CodeStriker<sup>[24]</sup>) designed specifically for
+code review) which often even included emailing around patchfiles.
+
+Finally, it is worth pointing out that many commercial companies, including Google,
+don't use formal branching mechanisms commonly available in revision control systems like Git
+to manage code review. Of course, they still engage in code review but do so by other means and
+instead of using branches keep everything merged on a single, monolithic line of development<sup>[25]</sup>
+(e.g. *trunk* or *mainline*). In portions of the code base in transition, both new and old
+code paths commonly exist simultaneously, controlled through the use of conditional flags,
+a practice that is highly conducive to incorporating large changes in small, incremental pieces.
```

Comment:
I am not sure this statement is entirely accurate.  I read the entire article https://cacm.acm.org/magazines/2016/7/204032-why-google-stores-billions-of-lines-of-code-in-a-single-repository.  While it seems true that many new features are developed side-by-side with existing features and [flags](https://martinfowler.com/articles/feature-toggles.html) are used to switch between the old and new implementation (and to facilitate A/B experiments), this article does not give any indication that most **refactorings** are done in this way.  (That is, the old code and the refactored code are not present in trunk at the same time.)  Instead it seems that a proposed sets of changes to a set of files are kept in different local "workspaces" against a specific "snapshot" version of trunk.  These "workspace snapshots" are the units of code reviews.  The article says:

> Files in a workspace are committed to the central repository only after going through the Google code-review process, as described later.

So while these "workspace snapshots" are in review, there may be changes being made to the same lines in the same files by other developers and development teams (and refactoring efforts). So the same problems of concurrent changes that are at the heart of the biggest problems of version control strategies remain.  And the longer that these "workspace snapshots" sit in review (and the more lines of code they touch), the greater the chance of (semantic or textural) conflicts will occur with what is being committed to trunk.  (Perhaps the Google system also has tools to detect conflicts between different "workspace snasphots" and updates to trunk and warn developers about them.  I would almost guarantee they have such tools and processes.  But being informed of conflicts and resolving conflicts are two different issues.)  I would really love to hear how Google addresses this core problem. (That article does not even mention this problem but we all know that it exists in all larger teams.)  They mention that most large refactorings are done first using smaller backward compatible sets of refactorings but that is almost impossible to do in some cases (like the change of an large API used by lots of clients).   (But other refactorings like fixing up code to work with newer C++ standards can be done very incrementally.) 

I guess the assumption is that the diffs in these "workspace snapshots" will generally be small and short-lived, but that is **not** always true.  In fact, this document says that automated refactorings of the huge main Google repo can result in changes sets involving "hundreds or thousands of files" across the entire repo.  And refactorings that touch hundreds and thousands of files can impact and 10x to 100x that number of lines of code.  This is more similar to the VTK-6 to VTK-8 refactoring described above.  And so we see the real issue here.  There is no way around large change-sets for some types of refactorings and breaking them up into 100s of smaller incremental refactorings likely does not help.  So do we not do these types of refactorings because they result in PRs that are too big to review?  Or is that the review of such changes are broken down in many smaller review units (i.e. by the teams that own the different parts of the code being refactored) or that the review process for large refactorings are conducted differently than reviews of new feature development?  This Google article seems to suggest that the review of these large refactorings are partitioned between many different developers, who are the owners of the code being refactored:

> Rosie splits patches along project directory lines, relying on the code-ownership hierarchy described earlier to send patches to the appropriate reviewers.

This BSSW article should make it clear that there is often a big difference between new feature development (that can usually be done in small increments and use [feature toggles](https://martinfowler.com/articles/feature-toggles.html) to enable or disable the new code) and refactorings that often can't be done effectively by adding new code leaving old code alone and therefore result in large patches. 

Therefore, I think this text:

> In portions of the code base in transition, both new and old
code paths commonly exist simultaneously, controlled through the use of conditional flags,
a practice that is highly conducive to incorporating large changes in small, incremental pieces.

can be made more accurate by changing it to something like:

> With most development, both new and old
code paths commonly exist simultaneously, controlled through the use of conditional flags,
a practice that is highly conducive to incorporating large changes in small, incremental pieces.